Add way to export a list of arphitgg graphs as a GIF animation

maxsutton opened the pull request:
This allows the user to create a gif from a list of arphitgg objects. 

I've implemented it using both magick; and magick with gifski. 

Using magick only (agg_slides_magick) doesn't require an additional package, and you can specify a maximum number of loops. The gif then stays on the final frame.

Gifski is generally faster to produce gifs and the .gif files are smaller. 

I've written tests to check these two functions and written a function to properly check a produced gif against the reference gif, frame by frame. However, this function does not check that the delay between the frames is the same. 

Concerns:
This does fail one test, because I haven't produced the reference gifs under Windows. Hopefully that's quick to remedy. 

I have set a default filename if none is supplied. You may wish to remove this functionality and return an error instead.

PS. This is my first ever PR, so please feel free to let me know how I can improve my PR format.

maxsutton commented:
Seems like the Travis CI doesn't like the gifski package. I installs fine on Windows, but needs a Rust installation on linux. 

Happy for that feature to be removed as it's just an alternative.

angusmoore commented:
This looks great, thanks. This is a great feature addition - and one I hadn't thought of.

A few broader comments - I'll provide a specific code review too when I get a chance, though by and large it looks good.

1. My benchmarks (windows) had imagemagick faster (5.5 seconds to create the example from the tests, versus 6.5 for gifski). For simplicity of code/keeping dependencies small I'd be inclined to lean to just using the imagemagick implementation unless there's a compelling reason to adopt both? (As you note, I find that imagemagick produces files about twice as large - is there a reason to worry about file size with GIFs? I genuinely don't know - both are small, but the examples I'm using are 4 frames, which are not very representative I suspect. If file size does become a problem, perhaps we could think about downsampling the images - the default PNGs produced are quite high resolution, perhaps more than necessary...).
2. Could you add documentation to the plotting-options vignette. Probably makes sense to be it's own heading - I don't think it fits under any of the other headings. At the end of the document seems like a sensible place. 
3. Could you add a sentence to NEWS.md as well.
4. Getting travis to build gifski should just require adding `cargo` to the packages part of `.travis.yml` (where the bit about libmagick is).

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+library(magick)
```
No need to add library calls in packages (I think it might even considered bad practice); rather, you just annotate functions with their namespace (as you've done elsewhere). This line can just be deleted

maxsutton replied:
Fixed in commit [90aba24](https://github.com/angusmoore/arphit/pull/293/commits/90aba24fc683a6bdddb438669085c5753bbaff33)

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+  }, gg_list)
+  on.exit(unlink(png_files))
+
+  info <- magick::image_info(image_read(png_files[1]))
```
image_read needs ``magick::` namespace annotation

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+  img_list <- lapply(seq_along(gg_list), function(idx, gg_list) {
+    file_name <- file.path(tempdir(), paste0("animate-", idx, ".png"))
+    agg_draw(gg_list[[idx]], file_name)
+    list(path = file_name, img = image_read(file_name))
```
image_read needs `magick::` annotation

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+#' @param gg_list List of gg objects
+#' @param filename Path to save .gif to
+#' @param delay Delay in seconds for each frame
+#' @param loop Whether to loop through frames
```
```
#' @param loop Whether to loop through frames. Setting to zero will loop infinitely (default).
```

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+#'
+#' Flick through ggs in a gif
+#'
+#' @param gg_list List of gg objects
```
probably `arphitgg objects` rather than `gg`?

maxsutton replied:
Fixed in commit [3389dea5](https://github.com/angusmoore/arphit/pull/293/commits/3389dea586472ede9b71af5c460441c927d52a01)

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+#' @param gg_list List of gg objects
+#' @param filename Path to save .gif to
+#' @param delay Delay in seconds for each frame
+#' @param loop Whether to loop through frames
```
#' @param loop Whether to loop through frames. Setting to zero will loop infinitely (default).

maxsutton replied:
Fixed in commit [3389dea5](https://github.com/angusmoore/arphit/pull/293/commits/3389dea586472ede9b71af5c460441c927d52a01)

angusmoore commented on R/animate.R:
```diff
@@ -0,0 +1,91 @@
+
+  magick::image_write(animation, path = filename)
+
+  filename
```
To be consistent with how agg_draw works, I wonder if it makes sense to _show_ the gif (rather than save it to `agg.gif`) if no filename is specified, and write to disk if one is. Unfortunately displaying gifs in R doesn't seem to be straightforward. Magick can do it, but only on X11 systems, so that's not much help. But magick does seem to be able to open a gif in a browser using ` magick::image_browse(foo, browser = NULL)` - that's probably a good enough option. Thoughts?

maxsutton replied:
That seems reasonable. I've implemented that in commit [3863f977](https://github.com/angusmoore/arphit/pull/293/commits/3863f9777781deb81e8f97a0dd31af0339b97439)

maxsutton commented:
- I've removed the gifski implementation, so that the code is simpler. 
- Thanks for catching the missing `imagemagick::` calls. Those should be fixed. 
- Documentation has been changed as suggested
- The gif is opened externally if `filename` is not supplied
- Added feature to NEWS
- Added a heading at the end of the plotting options vignette

angusmoore commented on tests/testthat/test-animations.R:
```diff
@@ -0,0 +1,23 @@
+      agg_line() + agg_ylim(-20, 20, 5)
+  })
+
+  gif_file <- agg_slides(gg_list, filename = tempfile("ani", fileext = ".gif"))
```
`agg_slides` doesn't return the filename anymore, so this test doesn't work. Could just assign the tempfile name to `gif_file` first and use that in both `check_gif` and `agg_slides(... filename = gif_file)`

maxsutton replied:
`agg_slides` still returns the filename if it is given one. I've left it that way because that's the return value of `imagemagick::image_write`.

So the test passes on my computer with that code. If it doesn't pass on your computer, happy to make the change.

angusmoore replied:
My mistake, you are absolutely right. I obviously mucked something stupid up when I ran this, sorry!

angusmoore commented:
This is looking excellent - the documentation is really very good.

Only a couple minor things left and then LGTM.

1. The gif files for the vignettes don't need to be included in the repo - they get created by the vignette builder. 
2. `animations-2.gif` isn't used anymore.
3. I've attached a windows reference for `animations-1`. I think that should get tests passing. (hopefully github hasn't re-encoded it or something...)
![animations-1](https://user-images.githubusercontent.com/18376371/54399168-50601100-4711-11e9-9185-d8f230c56dcf.gif)

angusmoore commented on tests/testthat/helper-test-check.R:
```diff
@@ -26,6 +26,43 @@ check_graph <- function(p, filename, max_distortion = 0.99) {
   }
 }
 
+check_gif <- function(gif, filename, max_distortion = 0.99) {
```
The current travis failure is spurious - it's just an artifact of how the code checks that there aren't any unused test reference files. Changing check_gif to accept the list of gg, rather than a filepath, and then saving the gif into the tempdir() (as is done by check_graph) _should_ (I hope) get around this.
```
check_gif <- function(gg_list, filename, max_distortion = 0.99) {
comp_location <- paste0(tempdir(), "/", filename, ".gif")
  agg_slides(gg_list, comp_location)
```

angusmoore commented on tests/testthat/helper-test-check.R:
```diff
@@ -26,6 +26,43 @@ check_graph <- function(p, filename, max_distortion = 0.99) {
+
+  if (file.exists(reference_loc) & file.exists(gif)) {
+    reference <- magick::image_read(reference_loc)
+    comparison <- magick::image_read(gif)
```
```
 comparison <- magick::image_read(comp_location)
```

(As above)

angusmoore commented on tests/testthat/test-animations.R:
```diff
@@ -0,0 +1,23 @@
+  })
+
+  gif_file <- agg_slides(gg_list, filename = tempfile("ani", fileext = ".gif"))
+  expect_true(check_gif(gif_file, "animations-1"))
```
```
expect_true(check_gif(gg_list, "animations-1"))
```

maxsutton commented:
Hopefully that change makes the Travis CI happy.

Tests passing on my end.
